### ISSUE TITLE
Custom event returns empty variant

### ISSUE THREAD

**Issue by NIPE-SYSTEMS** [open62541/open62541#1839]
I'm trying to send a custom event (event type `ns=2;i=1005`, for nodeset see below) from an open62541-server to an open62541-client. It seems that when I'm sending a populated event to the client, the client gets some empty variants instead of the populated data. It seems that this only effects fields of the custom event type. Fields of e.g. the `BaseEventType` seem to work fine.

I might have done it wrong. After hours of debugging I haven't found the issue.

The detailed informations below consist of (hopefully) the whole code that is needed for triggering and receiving the custom event:

1. open62541 client: subscribing to the event
2. open62541 client: subscription callback
3. open62541 server: triggering the event
4. and: the nodeset of the custom event type

### 1. open62541 client: subscribing to the event

```cpp
#define UA_CLIENT_SELECT_CLAUSES_AMOUNT 3

uint16_t m_namespace = 2; // via UA_Client_NamespaceGetIndex

UA_MonitoredItemCreateRequest request = UA_MonitoredItemCreateRequest_default(UA_NODEID_NUMERIC(0, UA_NS0ID_SERVER));
request.itemToMonitor.attributeId = UA_ATTRIBUTEID_EVENTNOTIFIER;
request.requestedParameters.samplingInterval = 100;
UA_EventFilter_init(&m_filter);
m_filter.selectClauses = (UA_SimpleAttributeOperand*)UA_Array_new(UA_CLIENT_SELECT_CLAUSES_AMOUNT, &UA_TYPES[UA_TYPES_SIMPLEATTRIBUTEOPERAND]);
if(!m_filter.selectClauses) {
    m_logger.error("Failed to allocate select clauses for subscription.");
    return false;
}
for(size_t i = 0; i < UA_CLIENT_SELECT_CLAUSES_AMOUNT; ++i) {
    UA_SimpleAttributeOperand_init(&m_filter.selectClauses[i]);
}

// ---------------------------------------------------------------------
// SELECT CLAUSE 1 ("2:IsAddition" from ns=2;i=1005)

m_filter.selectClauses[0].typeDefinitionId = UA_NODEID_NUMERIC(m_namespace, 1005);
m_filter.selectClauses[0].browsePathSize = 1;
m_filter.selectClauses[0].browsePath = (UA_QualifiedName*)UA_Array_new(m_filter.selectClauses[0].browsePathSize, &UA_TYPES[UA_TYPES_QUALIFIEDNAME]);
if(!m_filter.selectClauses[0].browsePath) {
    m_logger.error("Failed to allocate browsepath for subscription.");
    UA_SimpleAttributeOperand_delete(m_filter.selectClauses);
    return false;
}
m_filter.selectClauses[0].attributeId = UA_ATTRIBUTEID_VALUE;
m_filter.selectClauses[0].browsePath[0] = UA_QUALIFIEDNAME_ALLOC(m_namespace, "IsAddition");

// ---------------------------------------------------------------------
// SELECT CLAUSE 2 ("2:Element" from ns=2;i=1005)

m_filter.selectClauses[1].typeDefinitionId = UA_NODEID_NUMERIC(m_namespace, 1005);
m_filter.selectClauses[1].browsePathSize = 1;
m_filter.selectClauses[1].browsePath = (UA_QualifiedName*)UA_Array_new(m_filter.selectClauses[1].browsePathSize, &UA_TYPES[UA_TYPES_QUALIFIEDNAME]);
if(!m_filter.selectClauses[1].browsePath) {
    m_logger.error("Failed to allocate browsepath for subscription.");
    UA_SimpleAttributeOperand_delete(m_filter.selectClauses);
    return false;
}
m_filter.selectClauses[1].attributeId = UA_ATTRIBUTEID_VALUE;
m_filter.selectClauses[1].browsePath[0] = UA_QUALIFIEDNAME_ALLOC(m_namespace, "Element");

// ---------------------------------------------------------------------
// SELECT CLAUSE 3 ("0:Severity" from built-in base event type)

m_filter.selectClauses[2].typeDefinitionId = UA_NODEID_NUMERIC(0, UA_NS0ID_BASEEVENTTYPE);
m_filter.selectClauses[2].browsePathSize = 1;
m_filter.selectClauses[2].browsePath = (UA_QualifiedName*)UA_Array_new(m_filter.selectClauses[2].browsePathSize, &UA_TYPES[UA_TYPES_QUALIFIEDNAME]);
if(!m_filter.selectClauses[2].browsePath) {
    m_logger.error("Failed to allocate browsepath for subscription.");
    UA_SimpleAttributeOperand_delete(m_filter.selectClauses);
    return false;
}
m_filter.selectClauses[2].attributeId = UA_ATTRIBUTEID_VALUE;
m_filter.selectClauses[2].browsePath[0] = UA_QUALIFIEDNAME_ALLOC(0, "Severity");

// ---------------------------------------------------------------------
// triggering the event

m_filter.selectClausesSize = UA_CLIENT_SELECT_CLAUSES_AMOUNT;
request.requestedParameters.filter.encoding = UA_EXTENSIONOBJECT_DECODED;
request.requestedParameters.filter.content.decoded.data = &m_filter;
request.requestedParameters.filter.content.decoded.type = &UA_TYPES[UA_TYPES_EVENTFILTER];
UA_MonitoredItemCreateResult response = UA_Client_MonitoredItems_createEvent(m_client, m_subscriptionId, UA_TIMESTAMPSTORETURN_SOURCE, request, NULL, &UaClient::callback_subscriptionEvent, NULL);
if(response.statusCode != UA_STATUSCODE_GOOD) {
    m_logger.error("Failed to setup subscription for place.");
    return false;
}
m_monitoredItemId_event = response.monitoredItemId;
```

### 2. open62541 client: subscription callback

```cpp
void UaClient::callback_subscriptionEvent(UA_Client *_client, UA_UInt32 subId, void *subContext, UA_UInt32 monId, void *monContext, size_t nEventFields, UA_Variant *eventFields) {
    UaClient &client = UaClient::instance();
    bool setIsAddition = false; // flag whether isAddition has been set by event
    bool isAddition = false; // stores the isAddition field
    bool setElement = false; // flag whether element has been set by event
    uint32_t element = 0; // stores the element field
    client.m_logger.debug("subscriptionId: " + std::to_string(subId) + ", monitoringId: " + std::to_string(monId) + ", m_monitoredItemId_event: " + std::to_string(client.m_monitoredItemId_event));
    client.m_logger.debug("Got event with " + std::to_string(nEventFields) + " fields.");
    for(size_t i = 0; i < nEventFields; ++i) {
        if(UA_Variant_hasScalarType(&eventFields[i], &UA_TYPES[UA_TYPES_BOOLEAN])) {
            // this code never runs, because the corresponding eventField is an empty variant
            client.m_logger.debug("has scalar type: UA_TYPES_BOOLEAN");
            isAddition = *reinterpret_cast<bool *>(eventFields[i].data);
            setIsAddition = true;
            client.m_logger.info("IsAddition: " + std::to_string(isAddition));
        } else if(UA_Variant_hasScalarType(&eventFields[i], &UA_TYPES[UA_TYPES_UINT32])) {
            // this code never runs, because the corresponding eventField is an empty variant
            client.m_logger.debug("has scalar type: UA_TYPES_UINT32");
            element = *reinterpret_cast<uint32_t *>(eventFields[i].data);
            setElement = true;
            client.m_logger.info("Element: " + std::to_string(element));
        } else if(UA_Variant_isEmpty(&eventFields[i])) {
            client.m_logger.warn("Unknown type: sent variant is empty, ignoring.");
        } else {
            client.m_logger.warn("Unknown type: " + std::string(eventFields[i].type->typeName) + ", ignoring.");
        }
    }
    if(setIsAddition && setTool) {
        client.m_logger.info("<-- event: " + std::to_string(isAddition) + ", " + std::to_string(element));
    } else {
        client.m_logger.warn("<-- event: incomplete, ignoring.");
    }
}
```

The logging tells me:

```
2018-06-11 08:36:03.429 DEBUG client    subscriptionId: 1, monitoringId: 2, m_monitoredItemId_event: 2
2018-06-11 08:36:03.429 DEBUG client    Got event with 3 fields.
2018-06-11 08:36:03.429  WARN client    Unknown type: sent variant is empty, ignoring.
2018-06-11 08:36:03.429  WARN client    Unknown type: sent variant is empty, ignoring.
2018-06-11 08:36:03.429  WARN client    Unknown type: UInt16, ignoring.
2018-06-11 08:36:03.429  WARN client    <-- event: incomplete, ignoring.
```

So there are three fields sent via the event. One of them is an `uint16_t` which is the severity (yes, haven't printed that but I've proven that via my debugging, it is correct :wink:). The other two fields are empty which is strange.

### 3. open62541 server: triggering the event

```cpp
uint32_t id = 42;

UA_NodeId nodeIdEvent;
if(UA_Server_createEvent(server.m_server, UA_NODEID_NUMERIC(server.m_namespace, 1005), &nodeIdEvent) != UA_STATUSCODE_GOOD) {
    server.m_logger.warn("Failed to create event for elements addition.");
    return UA_STATUSCODE_BADINTERNALERROR;
}
UA_Boolean isAddition = UA_TRUE;
UA_UInt16 severity = 42;
UA_Server_writeObjectProperty_scalar(server.m_server, nodeIdEvent, UA_QUALIFIEDNAME_STATIC(server.m_namespace, "IsAddition"), &isAddition, &UA_TYPES[UA_TYPES_BOOLEAN]);
UA_Server_writeObjectProperty_scalar(server.m_server, nodeIdEvent, UA_QUALIFIEDNAME_STATIC(server.m_namespace, "Element"), &id, &UA_TYPES[UA_TYPES_UINT32]);
UA_Server_writeObjectProperty_scalar(server.m_server, nodeIdEvent, UA_QUALIFIEDNAME_STATIC(0, "Severity"), &severity, &UA_TYPES[UA_TYPES_UINT16]);
if(UA_Server_triggerEvent(server.m_server, nodeIdEvent, UA_NODEID_NUMERIC(0, UA_NS0ID_SERVER), NULL, UA_TRUE) != UA_STATUSCODE_GOOD) {
    server.m_logger.warn("Failed to trigger event for elements addition.");
    return UA_STATUSCODE_BADINTERNALERROR;
}
```

### 4. nodeset of the custom event type

```xml
<UAObjectType NodeId="ns=2;i=1005" BrowseName="2:ChangeEventType">
    <DisplayName Locale="en">ChangeEventType</DisplayName>
    <Description Locale="en">Event type that notifies about a change of the list of elements.</Description>
    <References>
        <Reference ReferenceType="HasSubtype" IsForward="false">i=2041</Reference>
        <Reference ReferenceType="HasComponent">ns=2;i=6011</Reference>
        <Reference ReferenceType="HasComponent">ns=2;i=6009</Reference>
        <Reference ReferenceType="GeneratesEvent" IsForward="false">ns=2;i=5001</Reference>
    </References>
</UAObjectType>
<UAVariable DataType="Boolean" ParentNodeId="ns=2;i=1005" NodeId="ns=2;i=6011" BrowseName="1:IsAddition" UserAccessLevel="3" AccessLevel="3">
    <DisplayName Locale="en">IsAddition</DisplayName>
    <Description Locale="en">True if this event corresponds to an addition, false if it corresponds to a removal.</Description>
    <References>
        <Reference ReferenceType="HasTypeDefinition">i=63</Reference>
        <Reference ReferenceType="HasComponent" IsForward="false">ns=2;i=1005</Reference>
        <Reference ReferenceType="HasModellingRule">i=78</Reference>
    </References>
    <Value>
        <uax:Boolean>false</uax:Boolean>
    </Value>
</UAVariable>
<UAVariable DataType="UInt32" ParentNodeId="ns=2;i=1005" NodeId="ns=2;i=6009" BrowseName="1:Element" UserAccessLevel="3" AccessLevel="3">
    <DisplayName Locale="en">Element</DisplayName>
    <Description Locale="en">The element that has been added/removed.</Description>
    <References>
        <Reference ReferenceType="HasTypeDefinition">i=63</Reference>
        <Reference ReferenceType="HasModellingRule">i=78</Reference>
        <Reference ReferenceType="HasComponent" IsForward="false">ns=2;i=1005</Reference>
    </References>
    <Value>
        <uax:UInt32>0</uax:UInt32>
    </Value>
</UAVariable>
```

I'm leaving out the logs of open62541 with `#define UA_LOGLEVEL 100`, there is nothing interesting, no errors. Just publish requests and the normal notification sending. Do you need Wireshark dumps?

You told me in https://github.com/open62541/open62541/issues/1824#issuecomment-393624451 that events are a bleeding edge feature. Is the feature still not usable? Am I too early again? :wink:

My open62541 HEAD is currently on fd304ae6452b398a49ff5028ae2458fbcfe8aadb.

Thanks in advance!

**Comment by aDogCalledSpot** [comment]
The null values occur when an EventFilter has fields which aren't covered by the event. We could fix this by skipping fields which don't apply and using a realloc at the end. Should not be a lot effort, but unfortunately, I'm hard pressed for time this week, so marking it as experimental might be a good idea.

**Comment by NIPE-SYSTEMS** [comment]
@aDogCalledSpot For me it would be interesting if I made a mistake in my code. Or do you need time to investigate this question? Maybe I can fix my code in parallel to the potential library-fix if you see a weird usage by my code. Thanks!

**Comment by jpfr** [comment]
On the current master branch, using UAExpert results in some empty fields.
But not all fields empty!

With UAExpert, generating events yields the following trace output:

```
13:23:59.028Z|4|33D0* [uastack] OpcUa_Channel_ResponseAvailable: CallResponse
13:23:59.061Z|4|33D0* [uastack] OpcUa_Channel_ResponseAvailable: PublishResponse
13:23:59.071Z|4|20C4* [uastack] OpcUa_Channel_BeginInvokeService: called for PublishRequest!
13:23:59.262Z|2|33D0* [uastack] OpcUa_Channel_ResponseAvailable: Decoding failed! (0x80B00000)
13:23:59.273Z|1|3010* Error: Publish failed [status=0x80b00000]
```

The error code translates to
```
#define UA_STATUSCODE_BADENDOFSTREAM 0x80B00000 /* Cannot move beyond end of the stream. */
```

**Comment by NIPE-SYSTEMS** [comment]
A moment ago, I discovered that I use `UA_Server_writeObjectProperty_scalar()` for writing the child nodes of the event node. In contrast, my `ChangeEventType` references the child nodes via `HasComponent`. The documentation of `UA_Server_writeObjectProperty_scalar()` states that it can modify child nodes that are referenced via `HasProperty`.

This might explain why the default fields of the event type (e.g. `Severity` via `HasProperty`) are populated but my special fields not (e.g. `IsAddition` via `HasComponent`).

I will let you know whether a corrected reference type fixes this issue.

**Comment by NIPE-SYSTEMS** [comment]
Question: Should I use the correct functions for writing OR adjust the reference types of my child fields?

`BaseEventType` contains several properties (`HasProperty`) but an event type of another major companion standard contains variables (`HasComponent`).

(In the meantime I will search for other companion standards to see how they are defining events. I will then update this comment.)

**Comment by NIPE-SYSTEMS** [comment]
@jpfr How are you using UaExpert to receive events? I've dragged the `Server` object into the `Event View` document and selected simply all events.

When I click on the `Refresh` icon button of the `Events` area, I get the following error:

```
Event Plugin: Call ConditionRefresh returned BadInternalError for ServerId Existing App.
```

I'm expecting events to appear in the `Events` area but there is nothing. :cry:

![image](https://user-images.githubusercontent.com/3900641/41501653-cdb25c56-71a8-11e8-8f66-9334fdd88065.png)

**Comment by aDogCalledSpot** [comment]
I have double-checked the specification on this and it states in Part 4, 7.17.3 (page 133):

> A null value is returned in the corresponding event field in the Publish response if the selected field is not part of the Event or an error was returned in the selectClauseResults of the EventFilterResult.

So I would say this behaviour is desired.

**Comment by NIPE-SYSTEMS** [comment]
Thanks @aDogCalledSpot, I agree with you, from the perspective of open62541 the leaving-fields-empty behaviour seems to be correct. In my scenario I mistakenly used components instead of properties for describing my event but filled the data in via properties.

Currently, I've not enough time to test the events (develop a open62541 client that receives these events). I'm okay with closing this issue (and reopen it myself if my server doesn't work).
